compiler/rustc_hir_typeck/src/method/probe.rs

```diff
@@ -950,51 +953,45 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
-                    this.xform_self_ty(&item, new_trait_ref.self_ty(), new_trait_ref.substs);
-                this.push_candidate(
-                    Candidate {
-                        xform_self_ty,
-                        xform_ret_ty,
-                        item,
-                        import_ids: import_ids.clone(),
-                        kind: TraitCandidate(new_trait_ref),
-                    },
-                    false,
-                );
-            });
-        } else {
-            debug_assert!(self.tcx.is_trait(trait_def_id));
-            if self.tcx.trait_is_auto(trait_def_id) {
-                return;
+        debug_assert!(self.tcx.is_trait(trait_def_id));
+        if self.tcx.trait_is_auto(trait_def_id) {
+            return;
+        }
+        for item in self.impl_or_trait_item(trait_def_id) {
+            // Check whether `trait_def_id` defines a method with suitable name.
+            if !self.has_applicable_self(&item) {
+                debug!("method has inapplicable self");
+                self.record_static_candidate(CandidateSource::Trait(trait_def_id));
+                continue;
             }
-            for item in self.impl_or_trait_item(trait_def_id) {
-                // Check whether `trait_def_id` defines a method with suitable name.
-                if !self.has_applicable_self(&item) {
-                    debug!("method has inapplicable self");
-                    self.record_static_candidate(CandidateSource::Trait(trait_def_id));
-                    continue;
-                }
 
-                let (xform_self_ty, xform_ret_ty) =
-                    self.xform_self_ty(&item, trait_ref.self_ty(), trait_substs);
-                self.push_candidate(
-                    Candidate {
-                        xform_self_ty,
-                        xform_ret_ty,
-                        item,
-                        import_ids: import_ids.clone(),
-                        kind: TraitCandidate(trait_ref),
-                    },
-                    false,
-                );
+            let (xform_self_ty, xform_ret_ty) =
+                self.xform_self_ty(&item, trait_ref.self_ty(), trait_substs);
+            self.push_candidate(
+                Candidate {
+                    xform_self_ty,
+                    xform_ret_ty,
+                    item,
+                    import_ids: import_ids.clone(),
+                    kind: TraitCandidate(trait_ref),
+                },
+                false,
+            );
+        }
+    }
+
+    fn assemble_extension_candidates_for_trait_or_alias(
+        &mut self,
+        import_ids: &SmallVec<[LocalDefId; 1]>,
+        trait_def_id: DefId,
+    ) {
+        debug!("assemble_extension_candidates_for_trait_or_alias(trait_def_id={:?})", trait_def_id);
+        if self.tcx.is_trait_alias(trait_def_id) {
+            for &bound in self.tcx.trait_alias_def(trait_def_id).explicit_bounds() {
```

Comment:
You can just use explicit_item_bounds and filter for trait predicates with Self as the self type here. No need to re-encode these bounds over again.

Reply:
How would that work across crates?  In particular if the trait alias is defined upstream?  `expect_item_bounds` requires the `DefId` to be local:

https://github.com/rust-lang/rust/blob/a00e24d76a9ac03b146d3bf89e2161b2466c6551/compiler/rustc_hir_analysis/src/collect/item_bounds.rs#L82-L86

Reply:
https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.explicit_item_bounds

This is a query that takes a DefId

Reply:
Actually, might need to use explicit_predicates_of or super_predicates_of, but the idea is the same. 

The point is that we just want the direct super predicates, and not the elaborated list.

Reply:
Makes sense, will do.  Thanks!